Backport several style changes from Gecko (4)

Loirooriol opened the pull request:
<!-- Please describe your changes on the following line: -->
This continues https://github.com/servo/servo/pull/29816.

---
<!-- Thank you for contributing to Servo! Please replace each `[ ]` by `[X]` when the step is complete, and replace `___` with appropriate data: -->
- [X] `./mach build -d` does not report any errors
- [X] `./mach test-tidy` does not report any errors
- [ ] These changes fix #___ (GitHub issue number if applicable)

<!-- Either: -->
- [ ] There are tests for these changes OR
- [ ] These changes do not require tests because ___

<!-- Also, please make sure that "Allow edits from maintainers" checkbox is checked, so that we can help you if you get stuck somewhere along the way.-->

<!-- Pull requests that do not address these steps are welcome, but they will require additional verification as part of the review process. -->

emilio committed:
style: Simplify :dir() implementation

This I noticed while working on the following patches. Shouldn't have
any behavior change: the behavior does in fact match the element state
flag semantics correctly if we do this. We did split the dir flags into
two element bits a while ago.

:not(:dir()) still behaves correctly of course, and we have tests for that.

Differential Revision: https://phabricator.services.mozilla.com/D130734

emilio committed:
style: Refactor :-moz-lwtheme pseudo-classes to get invalidated correctly

Use the same document state mechanism we have for :moz-locale-dir. Also,
simplify the setup of the later to be the same as :dir(), allowing the
matching code to be less repetitive.

This should fix some flakiness in chrome mochitests, but we have no existing
tests for these pseudo-classes more generally and since they're just
chrome-only I'm not super-excited about adding more.

Differential Revision: https://phabricator.services.mozilla.com/D130735

emilio committed:
style: Allow matches() / querySelector() / etc on chrome docs to access chrome-only selectors. r=boris

Without this some tests fail with the previous patch because code like:

  https://searchfox.org/mozilla-central/rev/267682a8f45221bf0bfe999d4a0239706a43bc56/browser/base/content/browser-gestureSupport.js#651

starts throwing. Unfortunately I had missed that on my try run, because
the error message didn't include that exception (it seemed like an
intermittent browser-chrome failure instead).

We could expose a ChromeOnly API for this, but this seems better. This
fixes it trivially, and also removes the "no url data" situation from
the selector parser, which is nice.

Differential Revision: https://phabricator.services.mozilla.com/D130818

emilio committed:
style: Unexpose :-moz-locale-dir and :-moz-lwtheme* from content

Behind a pref for now. Given these selectors do nothing on non-chrome
documents (they just don't match) it seems worth trying.

A cursory search seems to indicate they're not used for UA detection or
something like that (or at least I haven't found such an usage).

Differential Revision: https://phabricator.services.mozilla.com/D130736

emilio committed:
style: Remove some prefs for pseudo-classes that we have shipped/unshipped successfully

And remove code for :-moz-submit-invalid completely.

Differential Revision: https://phabricator.services.mozilla.com/D130737

emilio committed:
style: Add a foreground color to attention selection

In some platforms (like macOS, windows dark mode, android, and some gtk
themes) the foreground selection color might be `currentcolor`, and that
doesn't generally guarantee enough contrast with the attention
background.

Remove HeadlessLookAndFeelGTK's handling of this color since it's
useless (always overridden by prefs in all.js)

Differential Revision: https://phabricator.services.mozilla.com/D130617

Commit:
style: [css-lists] Style system changes to support 'reversed(<counter-name>)'

Differential Revision: https://phabricator.services.mozilla.com/D129955

Commit:
style: [css-lists] Remove the internal '-moz-list-reversed' CSS property that is no longer needed

Differential Revision: https://phabricator.services.mozilla.com/D129957

emilio committed:
style: Use preferred color scheme when forcing colors with system colors (except windows HCM)

This causes (among other things) pages to be dark when using regular
windows system colors and forcing colors to "always", which is nice.

Differential Revision: https://phabricator.services.mozilla.com/D131165

Loirooriol committed:
Further changes required by Servo

Loirooriol committed:
Further changes required by Servo

emilio committed:
style: Partially back out bug 1105364

Apparently using transparent borders and outlines is a common
accessibility technique to make those visible in HCM.

Bug 1740924 comment 9 seemed to indicate we were the only browser
rendering those browsers, but I just confirmed that Edge at least does
show them.

Keep respecting system colors as that's per spec.

Differential Revision: https://phabricator.services.mozilla.com/D131412

emilio committed:
style: Make custom properties that are IACVT guaranteed-invalid

This effectively backs out bug 1623396.

See:

  https://github.com/w3c/csswg-drafts/pull/6006
  https://drafts.csswg.org/css-variables/#guaranteed-invalid-value

And related discussion. Matches Chrome stable as per https://groups.google.com/a/chromium.org/g/blink-dev/c/0xrbzYe_vxU/m/7bsL76n9CgAJ

Depends on D116459

Differential Revision: https://phabricator.services.mozilla.com/D116460

aethanyc committed:
style: Run rustfmt on servo/components/style and servo/ports/geckolib

This patch is generated by running `cargo +nightly fmt` under
`servo/components/style/` and `servo/ports/geckolib` against mozilla-central
https://hg.mozilla.org/mozilla-central/rev/b193f2e7a6a5d1f042c957ea4acd5c89bf210512

My nightly version is: 1.58.0-nightly (c9c4b5d72 2021-11-17)

Manually remove the redundant braces in author_styles.rs to fix a warning.

Differential Revision: https://phabricator.services.mozilla.com/D131556

aethanyc committed:
style: Support scrollbar-gutter in the style system

This patch adds `scrollbar-gutter` property in CSS Overflow level 3 [1] to the
style system.

`devtools/shared/css/generated/properties-db.js` is generated by
`./mach devtools-css-db`.

[1] https://drafts.csswg.org/css-overflow-3/#scrollbar-gutter-property

Differential Revision: https://phabricator.services.mozilla.com/D131460

emilio committed:
style: Top layer elements that are display:contents should be display: block

As per https://fullscreen.spec.whatwg.org/#new-stacking-layer:

    If its specified display property is contents, it computes to block.

Differential Revision: https://phabricator.services.mozilla.com/D131585

emilio committed:
style: Use first generic rather than only generic to determine fallback font family

This seems like more sensible behavior. We have another use of
only_generic(), but that affects font sizing and other browsers agree
with us there:

  <div style="font-family: monospace">Should be 13px</div>
  <div style="font-family: something, monospace">Should be 16px</div>

So not touching that one.

Differential Revision: https://phabricator.services.mozilla.com/D130732

Loirooriol committed:
Further changes required by Servo

emilio committed:
style: Simplify language-dependent font fallback code

.fallback is always the default font for the lang group unless we're a
system font (in which case it's "none").

The only reason we need that is because we need to react to
language changes (which affect the initial font).

Simplify the model a bit doing the language lookup in gfxTextRun (this
should avoid allocating a few nsStyleFont structs too.

Depends on D130732

Differential Revision: https://phabricator.services.mozilla.com/D131312

lamoure6 committed:
style: Removed remaining (non-negated) -moz-proton media queries

Differential Revision: https://phabricator.services.mozilla.com/D125328

emilio committed:
style: Correctly report animation status of pseudo-elements that are not stored in the parent element

We allow animating pseudo-elements like ::-moz-progress-bar (and we
treat them like regular elements).

Ideally we should store animations for these in the parent element as
well, so they survive reframes and such. But treating them as regular
elements right now means that we do animate them, but we never update
animations for them correctly because wrapper.rs assumed them to be
non-animatable.

Since it seems reasonable to keep allowing the animations to happen,
let's just correct the update code and add a test.

Differential Revision: https://phabricator.services.mozilla.com/D131794

emilio committed:
style: Keep system fonts when prioritizing user fonts

This was a subtle behavior change in bug 1739464. The early-return here:

  https://hg.mozilla.org/mozilla-central/rev/2fb74c67b0c5#l11.40

Meant we also bailed out from user font prioritization for system fonts.

It's unclear whether that's really the best behavior but since the
intention of the regressing patch was not to change behavior, preserving
the old behavior seems better.

Differential Revision: https://phabricator.services.mozilla.com/D131899

emilio committed:
style: Add a Show Password button to <input type=password> controls

It's controlled by the pref:
layout.forms.input-type-show-password-button.enabled

Differential Revision: https://phabricator.services.mozilla.com/D130407

emilio committed:
style: Honor GTK button layout

This is based off work by smurfd. But this patch doesn't support buttons
both at the left and right, which simplifies a lot the implementation.

Also, clean-up the existing env variables while at it.

Co-authored-by: Nicklas Boman <smurfd@gmail.com>

Differential Revision: https://phabricator.services.mozilla.com/D132073

emilio committed:
style: Remove unnecessary button-focus appearance value on Linux

This always draws transparent. Only real thing it does is forcing 0px
padding.

Differential Revision: https://phabricator.services.mozilla.com/D132563

emilio committed:
style: Use titlebar radius on Linux and make titlebar set-up work for lightweight themes

To do this, we always draw the native titlebar behind the toolbox, and
then make the toolbox adapt to it by using the titlebar radius. This
makes us preserve the shadow properly.

On Wayland we'd double-draw the shadow (see bug 1509931 comment 4) so
this fixes it by trimming it as well using border-radius.

Differential Revision: https://phabricator.services.mozilla.com/D128681

Loirooriol committed:
Further changes required by Servo

AlaskanEmily committed:
style: Implement parsing of the page property

Differential Revision: https://phabricator.services.mozilla.com/D131531

BorisChiou committed:
style: Part 1: Add ScrollTimeline class

Define a simple version of ScrollTimeline.

Differential Revision: https://phabricator.services.mozilla.com/D129099

BorisChiou committed:
style: Part 8: Hook scroll-timeline rule into Cascade data and use it for CSS animations

We hook the rule into cascade data, and so we can look it up by timeline
name. Now we only use StyleScrollDirection from @scroll-timeline rule.
`source` and `scroll-offsets` are skipped now and use the default values
instead because I'm pretty sure the syntax will be changed in Bug 1733260,
and `scroll-offsets` may be obsolete because the spec proposal intents to
make it be always 0% ~ 100%.

Also, add some reftests for the default `source` and `scroll-offsets`,
and different `orientation`s.

Besides, we disable at-scroll-timeline-start-end.html in Gecko because
we don't support start/end descriptors, and there are too many
intermittents in it.

Differential Revision: https://phabricator.services.mozilla.com/D126452

BorisChiou committed:
style: Part 10: Make source and scroll-offsets accept only default value

Based on our previous patches, we only support default behavior for
source and scroll-offsets:
1. source:auto
2. scroll-offsets: none
3. scroll-offsets: auto, auto, ...

So update the parser for them. We expect to remove whole
@scroll-timeline in Bug 1733260, so now only do a tiny update in parser.

Differential Revision: https://phabricator.services.mozilla.com/D132417

jfkthame committed:
style: Add support for the 'ic' font-relative unit

This is a "simplified" implementation of 'ic', similar to what Safari Preview
currently supports: it only considers the advance of U+6C34 if found in the
first available font, and otherwise falls back to the default of 1em.

(The spec allows for this "in cases where it is impossible or impractical to
determine the ideographic advance measure".)

Differential Revision: https://phabricator.services.mozilla.com/D132818

emilio committed:
style: Add support for the revert-layer keyword

This patch looks bigger than it is, but it's mostly because
of plumbing.

To implement revert-layer we need not only the cascade origin of the
declaration, but the whole cascade level, plus also the layer order.

In order to do this, encapsulate these two things inside a 32-bit
`CascadePriority` struct and plumb it through the rule tree and so on.
This allows us to remove the packing and unpacking of CascadeLevel,
though I kept the ShadowCascadeOrder limit for now in case we need to
reintroduce it.

Fix `!important` behavior of layers while at it (implementing it in
`CascadeLevel::cmp`, spec quote included since it was tricky to find)
since some revert-layer tests were depending on it.

The style attribute test is failing now, but follow-up commit fixes
it, see spec issue.

In terms of the actual keyword implementation, it's sort of
straight-forward: We implement revert and revert-layer in a shared
way, by storing the cascade priority that reverted it.

Differential Revision: https://phabricator.services.mozilla.com/D133372

Loirooriol committed:
Further changes required by Servo

emilio committed:
style: Fix style attribute important and revert-layer behavior

By modeling it as a separate layer that behaves somewhat specially.

See https://github.com/w3c/csswg-drafts/issues/6872.

The remaining revert-layer tests that we fail are because either we
don't implement a feature (like @property) or because it's used in
keyframes (where revert is a bit unspecified and we have existing
issues with it).

Differential Revision: https://phabricator.services.mozilla.com/D133373

emilio committed:
style: Update CSSOM for layer rules to the spec

Pretty mechanical.

Tests are in https://wpt.live/css/css-cascade/layer-rules-cssom.html
which (with a fix for @import tests which I'll submit separately) we
pass.

Sync for that is bug 1743936.

Differential Revision: https://phabricator.services.mozilla.com/D133387

emilio committed:
style: Don't draw window decorations when painting headerbar on wayland

This is a better fix for the double decorations than clipping them using
CSS.

Differential Revision: https://phabricator.services.mozilla.com/D133871

jfkthame committed:
style: Add CSS support for the hyphenate-character property

Differential Revision: https://phabricator.services.mozilla.com/D133889

emilio committed:
style: Deal with layers and at-rules

Differential Revision: https://phabricator.services.mozilla.com/D134010

emilio committed:
style: Remove servo/components/{hashglobe,fallible} in favor of try_reserve

Differential Revision: https://phabricator.services.mozilla.com/D134194

Loirooriol committed:
Further changes required by Servo

Loirooriol committed:
Further changes required by Servo

Loirooriol committed:
Further changes required by Servo

emilio committed:
style: Add hwb/a to devtools autocomplete lists

The changes to the devtools directory were written automatically via ./mach devtools-css-db

Differential Revision: https://phabricator.services.mozilla.com/D134197

emilio committed:
style: Remove hwba since it's not a thing (hwb function supports alpha)

Partially backs out the previous patch.

Differential Revision: https://phabricator.services.mozilla.com/D134256

emilio committed:
style: Use ThreadPool::scope_fifo in style

It does the same, but it saves an indentation level:

  https://searchfox.org/mozilla-central/rev/a11b63915bd7810a03635d733123448ab5bfcad3/third_party/rust/rayon-core/src/thread_pool/mod.rs#217

Differential Revision: https://phabricator.services.mozilla.com/D134321

emilio committed:
style: Remove ThinBoxedSlice

The only remaining consumers are ::-moz-tree pseudo-elements (we used to
use ThinBoxedSlice for other data structures in the past).

Those are not particularly performance sensitive so I think just
double-boxing is fine. In the future, if we wanted to avoid the double
indirection, we could probably use the "thin" crate
(https://docs.rs/thin) or similar, which stores the length of the slice
along with the allocation, making the pointer thin in all
configurations, much like "ThinArc" does:

  https://searchfox.org/mozilla-central/rev/1ce2eea39442190a71a1f8f650d098f286bf4a01/servo/components/servo_arc/lib.rs#891

In practice though, I don't think it's particularly worth it for this
specific case.

Differential Revision: https://phabricator.services.mozilla.com/D134672

emilio committed:
style: Shrink maps if needed after stylist rebuilds

Hashbrown grows a lot sometimes making us waste a lot of memory. Shrink
some of these maps after CascadeData rebuild / stylesheet collection
invalidation.

Differential Revision: https://phabricator.services.mozilla.com/D134716

emilio committed:
style: Make color-adjust an alias of print-color-adjust as per spec

Differential Revision: https://phabricator.services.mozilla.com/D134779

Commit:
style: Replace MathML font constants with enum

Differential Revision: https://phabricator.services.mozilla.com/D134802

emilio committed:
style: Rename StyleMathMLMathVariant -> StyleMathVariant

MANUAL PUSH: Trivial rename.

emilio committed:
style: Remove touch-action pref

It's been enabled by default since ~forever.

Differential Revision: https://phabricator.services.mozilla.com/D134935

emilio committed:
style: Avoid generating InterpolateMatrix operations if there are no size dependencies

The issue here is that we end up with a transition between mismatched
transform lists that ends up generating an InterpolateMatrix {}
operation. So far so good, but we end up interpolating that a lot of
times and generating an unboundedly-deep operation list.

This implementas an optimization that flattens them to a single matrix
when possible (when there's no dependencies on the containing box).

This is similar to:

  https://chromium.googlesource.com/chromium/src.git/+/2b89cc4df436e672ef9cf940d1c0dc73fef82a4a

We fix the to_pixel_length() behavior for LenghtPercentage to be
correct (and update callers to preserve behavior).

Differential Revision: https://phabricator.services.mozilla.com/D134784

emilio committed:
style: Apply line-height to ::marker

Differential Revision: https://phabricator.services.mozilla.com/D136313

emilio committed:
style: Choose tabpanel background based on content preferred color-scheme

If the theme is dark but user prefers light pages, the background of the
tabpanel should arguably be light, since it can be seen across some
navigations.

Expose a -moz-content-prefers-color-scheme media query to chrome pages
so that our UI can correctly query it (and remove the unused -moz-proton
atom while at it).

Differential Revision: https://phabricator.services.mozilla.com/D136437

AlaskanEmily committed:
style: Implement basic @page-rule selector parsing

This does not support any of the pseudo page types.

Differential Revision: https://phabricator.services.mozilla.com/D131532

AlaskanEmily committed:
style: Add style_traits::ToCss for AtomIdent

Differential Revision: https://phabricator.services.mozilla.com/D136290

glandium committed:
style: Remove MOZ_DIST

As mentioned in bug 1747354, the location of the dist directory is
relied to be $topobjdir/dist, so just use that consistently rather
than getting it from a separate variable for rust build scripts.

Differential Revision: https://phabricator.services.mozilla.com/D136556

glandium committed:
style: Use the mozbuild crate in servo

Differential Revision: https://phabricator.services.mozilla.com/D136562

emilio committed:
style: Style system and plumbing for mix-blend-mode: plus-lighter

Differential Revision: https://phabricator.services.mozilla.com/D137951

emilio committed:
style: Generalize moz-os-version into moz-platform

Allow differentiating non-windows platforms on it.

Differential Revision: https://phabricator.services.mozilla.com/D138431

emilio committed:
style: Make custom-property substitution only traverse variables which could have references

If a name is not in self.seen, it means we've inherited it from our
parent. That in turn means that it can't have any variable reference
(because we inherit the computed variables) and we can skip the work of
traversing it, as we'd hit the early-return in traverse() anyways.

This doesn't fix the memory usage issue of the page on the bug, which
has a giant list of properties on the root and then a custom property
specified on all elements, but should significantly reduce the time we
spend iterating over custom properties for all those elements.

Differential Revision: https://phabricator.services.mozilla.com/D140825

emilio committed:
style: Reuse inherited custom properties if they didn't change after resolution

This should be cheap and gives us a lot of memory savings for the page
on the bug, by deduplicating the inherited properties between parent and
children.

WebKit implements a similar optimization.

Differential Revision: https://phabricator.services.mozilla.com/D140826

mrobinson committed:
style: Add support for parsing the `content-visibility` property from the CSS Contain specification

Add initial parsing support for the CSS `content-visibility` attribute.
Currently these parsed values have no effect.

Differential Revision: https://phabricator.services.mozilla.com/D140834

emilio committed:
style: Honor background-color: transparent in forced colors mode

See comment as for why, and linked bugs, in particular:

  https://bugzilla.mozilla.org/show_bug.cgi?id=1755713#c16

And the following screenshot for example.

Differential Revision: https://phabricator.services.mozilla.com/D141514

emilio committed:
style: Put overflow: -moz-hidden-unscrollable behind a pref on Nightly

Differential Revision: https://phabricator.services.mozilla.com/D141759

emilio committed:
style: Use atomic ops to read / write node flags from stylo

The flags stylo cares about reading and writing potentially at the same
time are disjoint, so there's no need for any strong memory ordering.

Differential Revision: https://phabricator.services.mozilla.com/D141829

emilio committed:
style: Remove some no longer used use statements

MANUAL PUSH: Bustage fix CLOSED TREE

Loirooriol committed:
Further changes required by Servo

Loirooriol committed:
Further changes required by Servo

Loirooriol committed:
Further changes required by Servo

Loirooriol committed:
Further changes required by Servo

Loirooriol committed:
Further changes required by Servo

Loirooriol committed:
Further changes required by Servo

Loirooriol committed:
Further changes required by Servo

Loirooriol committed:
Avoid complaints from ./mach test-tidy

Loirooriol commented:
@bors-servo try=wpt

bors-servo commented:
:hourglass: Trying commit 144839beba277c7cab28d501d17a6bc65ac06a46 with merge 724f8e8595e736ada4522da4096ad9d3264c9e51...

bors-servo commented:
:broken_heart: Test failed - [checks-github](https://github.com/servo/servo/actions/runs/5194557113/jobs/9367120130)

Loirooriol committed:
Further changes required by Servo

emilio committed:
style: Make #[css(field_bound)] and #[css(iterable)] work properly

For now, use IntoIterator to figure the right type to add the bound.

If we need this on types that are iterable but don't provide
IntoIterator, we can add another attribute field or something.

Differential Revision: https://phabricator.services.mozilla.com/D129962

Loirooriol commented:
@bors-servo try=wpt

bors-servo commented:
:hourglass: Trying commit 2c1604fe0d64bf9e2361ca0da087c0c0cd1a5ee3 with merge 1922eecbb7ea2e563309cba9fc0a7b04fd9ffede...

Loirooriol committed:
Update test expectations

Commit:
style: Make #[css(represents_keyword)] convert underscore in the field name to dash

Differential Revision: https://phabricator.services.mozilla.com/D128668

Loirooriol committed:
Upgrade darling to 0.13.1

Based on https://phabricator.services.mozilla.com/D136568

Loirooriol commented:
@bors-servo try=wpt

bors-servo commented:
:hourglass: Trying commit d8b2d4b8d34cbe0b17f8319aeddf7194b6bad323 with merge 38ffc4f709159eca3c17cabd8666aed2f93a40e5...

Loirooriol commented:
@bors-servo try=wpt

bors-servo commented:
:hourglass: Trying commit 546d81ae818a2becd06682f56febe6076b956050 with merge 6131eaf72ad908ce91cd9bf71d06d754b636a853...

bors-servo commented:
:sunny: Test successful - [checks-github](https://github.com/servo/servo/actions/runs/5220343325/jobs/9424495257)
State: approved= try=True

Loirooriol commented:
@bors-servo try=wpt-2020

bors-servo commented:
:hourglass: Trying commit aefaa3f16cd3d8e2df7399f0d10022cd2cebd5a0 with merge 12e0fab791a1f61622a0d7479a084bee98fc1908...

bors-servo commented:
:sunny: Test successful - [checks-github](https://github.com/servo/servo/actions/runs/5221161981/jobs/9425865956)
State: approved= try=True

Loirooriol commented:
These are my patches: 07dbd9d63 7797a19c7 f911fb4f0 1d18b3a02 a95a74259 5c0f044d3 1ac55889b 98b49f802 07d1bd560 89041ac33 f9610e589 01549f038 e4bb1df87 f39ab4ffc 54878595b ebfbd2d9c e2901169c c4dbec650 aefaa3f16

mrobinson commented:
@bors-servo r+

bors-servo commented:
:pushpin: Commit aefaa3f has been approved by `mrobinson`

<!-- @bors-servo r=mrobinson aefaa3f16cd3d8e2df7399f0d10022cd2cebd5a0 -->

bors-servo commented:
:hourglass: Testing commit aefaa3f16cd3d8e2df7399f0d10022cd2cebd5a0 with merge 9260683b5fbc7fa9fe2cc75bbfccbcd9d0e245fd...

bors-servo commented:
:sunny: Test successful - [checks-github](https://github.com/servo/servo/actions/runs/5242287673/jobs/9468188987)
Approved by: mrobinson
Pushing 9260683b5fbc7fa9fe2cc75bbfccbcd9d0e245fd to master...
